Mathlib/Analysis/Asymptotics/ExpGrowth.lean

```diff
@@ -0,0 +1,319 @@
+
+lemma expGrowthSup_inv {u : ℕ → ℝ≥0∞} :
+    expGrowthSup u⁻¹ = - expGrowthInf u := by
+  rw [expGrowthInf, ← limsup_neg]
+  refine limsup_congr (Eventually.of_forall fun n ↦ ?_)
+  rw [Pi.neg_apply, Pi.inv_apply, div_eq_mul_inv, div_eq_mul_inv, ← EReal.neg_mul, log_inv]
+
+/-! ### Comparison -/
+
+-- Bound on `expGrowthInf` under a `IsBigO` hypothesis. However, `ℝ≥0∞` is not normed, so the
+-- `IsBigO` property is spelt out.
+lemma expGrowthInf_le_of_eventually_le {u v : ℕ → ℝ≥0∞} {C : ℝ≥0∞} (hC : C ≠ ∞)
+    (h : ∀ᶠ n in atTop, u n ≤ C * v n) :
+    expGrowthInf u ≤ expGrowthInf v := by
+  apply (expGrowthInf_eventually_monotone h).trans
+  rcases eq_zero_or_pos C with rfl | C_pos
+  · simp only [zero_mul, ← Pi.zero_def, expGrowthInf_zero, bot_le]
+  · apply (expGrowthInf_mul_le _ _).trans_eq <;> rw [expGrowthSup_const C_pos.ne' hC]
+    · exact zero_add (expGrowthInf v)
+    · exact Or.inl EReal.zero_ne_bot
+    · exact Or.inl EReal.zero_ne_top
+
+-- Bound on `expGrowthSup` under a `IsBigO` hypothesis. However, `ℝ≥0∞` is not normed, so the
+-- `IsBigO` property is spelt out.
+lemma expGrowthSup_le_of_eventually_le {u v : ℕ → ℝ≥0∞} {C : ℝ≥0∞} (hC : C ≠ ∞)
+    (h : ∀ᶠ n in atTop, u n ≤ C * v n) :
+    expGrowthSup u ≤ expGrowthSup v := by
+  apply (expGrowthSup_eventually_monotone h).trans
+  rcases eq_zero_or_pos C with rfl | C_pos
+  · simp only [zero_mul, ← Pi.zero_def, expGrowthSup_zero, bot_le]
+  · apply (expGrowthSup_mul_le _ _).trans_eq <;> rw [expGrowthSup_const C_pos.ne' hC]
+    · exact zero_add (expGrowthSup v)
+    · exact Or.inl EReal.zero_ne_bot
+    · exact Or.inl EReal.zero_ne_top
+
+lemma expGrowthInf_of_eventually_ge {u v : ℕ → ℝ≥0∞} {c : ℝ≥0∞} (hc : c ≠ 0)
+    (h : ∀ᶠ n in atTop, c * u n ≤ v n) :
+    expGrowthInf u ≤ expGrowthInf v := by
+  apply (expGrowthInf_eventually_monotone h).trans' (le_expGrowthInf_mul.trans' _)
+  rcases eq_top_or_lt_top c with rfl | c_top
+  · rw [← Pi.top_def, expGrowthInf_top]
+    exact le_add_of_nonneg_left le_top
+  · rw [expGrowthInf_const hc c_top.ne, zero_add]
+
+lemma expGrowthSup_of_eventually_ge {u v : ℕ → ℝ≥0∞} {c : ℝ≥0∞} (hc : c ≠ 0)
+    (h : ∀ᶠ n in atTop, c * u n ≤ v n) :
+    expGrowthSup u ≤ expGrowthSup v := by
+  apply (expGrowthSup_eventually_monotone h).trans' (le_expGrowthSup_mul'.trans' _)
+  rcases eq_top_or_lt_top c with rfl | c_top
+  · exact expGrowthInf_top ▸ le_add_of_nonneg_left le_top
+  · rw [expGrowthInf_const hc c_top.ne, zero_add]
+
+/-! ### Infimum and supremum -/
+
+lemma expGrowthInf_inf {u v : ℕ → ℝ≥0∞} :
+    expGrowthInf (u ⊓ v) = expGrowthInf u ⊓ expGrowthInf v := by
+  rw [expGrowthInf, expGrowthInf, expGrowthInf, ← liminf_min]
+  refine liminf_congr (Eventually.of_forall fun n ↦ ?_)
+  rw [Pi.inf_apply, log_monotone.map_min]
+  exact (monotone_div_right_of_nonneg n.cast_nonneg').map_min
+
+/-- Lower exponential growth as an `InfTopHom`.-/
+noncomputable def expGrowthInf_infTopHom : InfTopHom (ℕ → ℝ≥0∞) EReal where
```

Comment:
Since it's a def, there should be no underscore in it. Would `expGrowthInfHom` be too ambiguous for your taste?